runtime/compiler/codegen/J9CodeGenerator.cpp

```diff
@@ -156,6 +154,172 @@ static TR::Node *lowerCASValues(
    }
 
 
+/**
+ * @brief Add checks to skip (fast-path) acmpHelper call
+ * 
+ * @details
+ * 
+ * This transformation adds checks for the cases where the acmp can be performed
+ * without calling the VM helper. The trasformed Trees represen the following operation:
+ *
+ * 1. If the address of lhs and rhs are the same, produce an eq (true) result
+ *    and skip the call (note the two objects must be the same regardless of
+ *    whether they are value types are reference types)
+ * 2. Otherwise, do VM helper call
+ *
+ * The transformation looks as follows:
+ *
+ *  +----------------------+
+ *  |ttprev                |
+ *  |treetop               |
+ *  |  icall acmpHelper    |
+ *  |    aload lhs         |
+ *  |    aload rhs         |
+ *  |ificmpeq --> ...      |
+ *  |  ==> icall           |
+ *  |  iconst 0            |
+ *  |BBEnd                 |
+ *  +----------------------+
+ *
+ *  ...becomes...
+ *
+ *  +----------------------+
+ *  |ttprev                |
+ *  |iRegStore x           |
+ *  |  iconst 1            |
+ *  |ifacmpeq  -->---------*---------+
+ *  |  aload lhs           |         |
+ *  |  aload rhs           |         |
+ *  |BBEnd                 |         |
+ *  +----------------------+         |
+ *  |BBStart (extension)   |         |
```

Comment:
Whether this should be an extension and whether this should have the same frequency as the original block depends on how likely we think the we are to run the helper (if it is unlikely having this code down the bottom of the method might be more performant?)

Reply:
Agreed.

I was having trouble setting things up so the call could be in its own block, let alone putting towards the end of the function. Making it an extension was the easiest way I found of splitting the block without getting lost in the details of setting the proper global register dependencies everywhere.

Reply:
I'm fine with this for a first implementation but there should be an issue to track going back to fix this since it will become a performance problem in the long run.

Reply:
Opened #9857.